cert-generation/webhooks/webhook.go

```diff
@@ -0,0 +1,72 @@
+//
+// Copyright (c) 2019-2020 Red Hat, Inc.
+// This program and the accompanying materials are made
+// available under the terms of the Eclipse Public License 2.0
+// which is available at https://www.eclipse.org/legal/epl-2.0/
+//
+// SPDX-License-Identifier: EPL-2.0
+//
+// Contributors:
+//   Red Hat, Inc. - initial API and implementation
+//
+
+package webhooks
+
+import (
+	"github.com/devfile/devworkspace-operator/cert-generation/config"
+	v1beta1 "k8s.io/api/admissionregistration/v1beta1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+// WebhookInit Initialize the webhook that denies all everything until controller is started succesfully
+func WebhookInit(clientset *kubernetes.Clientset, namespace string) error {
+	mutateWebhookPath := config.MutateWebhookPath
+	webhookClientConfig := v1beta1.WebhookClientConfig{
+		Service: &v1beta1.ServiceReference{
+			Name:      config.SecureServiceName,
+			Namespace: namespace,
+			Path:      &mutateWebhookPath,
+		},
+	}
+
+	failPolicy := v1beta1.Fail
+	workspaceObjMutateWebhook := v1beta1.MutatingWebhook{
+		Name:          config.MutateWebhookName,
+		FailurePolicy: &failPolicy,
+		ClientConfig:  webhookClientConfig,
+		Rules: []v1beta1.RuleWithOperations{
+			{
+				Operations: []v1beta1.OperationType{v1beta1.Create},
+				Rule: v1beta1.Rule{
+					APIGroups:   []string{"workspace.devfile.io"},
+					APIVersions: []string{"v1alpha1"},
+					Resources:   []string{"*"},
+				},
+			},
+		},
+	}
+	configuration := &v1beta1.MutatingWebhookConfiguration{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: config.MutateWebhookCfgName,
+		},
+		Webhooks: []v1beta1.MutatingWebhook{
+			workspaceObjMutateWebhook,
+		},
+	}
+
+	// Create dummy mutating webhook or update it if it already exists
+	if _, err := clientset.AdmissionregistrationV1beta1().MutatingWebhookConfigurations().Get(config.MutateWebhookName, metav1.GetOptions{}); apierrors.IsNotFound(err) {
+		_, err = clientset.AdmissionregistrationV1beta1().MutatingWebhookConfigurations().Create(configuration)
+		if err != nil {
+			return err
+		}
+	} else {
+		_, err := clientset.AdmissionregistrationV1beta1().MutatingWebhookConfigurations().Update(configuration)
```

Comment:
I don't know if it's required to update the webhook if it already exists. By default the full webhook's failure policy is Fail, so once the regular webhook is created, a controller update will cause the webhooks to fail until the controller is running.

Reply:
Initially, my idea was to create stub mutating webhook configuration (if it's missing) that will be updated by controller. It would simplify workflow a bit I think.
Currently, we can imagine that cert-generation deployment is restarted for some reason (e.g. due to node restart) and that blocker webhook will be recreated and the controller will clean it but only when its next restart.